pkg/learning/mab.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package learning
+
+import (
+	"math"
+	"math/rand"
+	"sort"
+	"sync"
+
+	"github.com/google/syzkaller/pkg/log"
+)
+
+const MABLogLevel = 4
+
+type Choice struct {
+	reward     float64 // Normalized reward.
+	weight     float64 // Weight proportional to reward. Subject to float64 overflow prevention.
+	sumWeights float64 // Sum of weights up to current choice.
+}
+
+type MultiArmedBandit struct {
+	Mu      sync.RWMutex
+	Choices []Choice // Choices.
+
+	Theta float64 // Factor controlling gamma and eta.
+	Gamma float64 // Exploration factor.
+	Eta   float64 // Growth factor.
+}
+
+// Randomly chooses an "arm" in the Choices array.
+// Returns both the choice as well as the probability of making this choice.
+func (mab *MultiArmedBandit) Choose(r *rand.Rand) (int, float64) {
+	mab.Mu.RLock()
+	defer mab.Mu.RUnlock()
+
+	if len(mab.Choices) == 0 {
+		return -1, 0.0
+	}
+	sumWeights := mab.Choices[len(mab.Choices)-1].sumWeights
+	randVal := r.Float64() * mab.Choices[len(mab.Choices)-1].sumWeights
+	idx := sort.Search(len(mab.Choices), func(i int) bool {
+		return mab.Choices[i].sumWeights >= randVal
+	})
+	return idx, mab.Choices[idx].weight / sumWeights
+}
+
+func (mab *MultiArmedBandit) NewChoice() int {
```

Comment:
It's actually used in the test where the index of the new choice is checked. Admittedly, we could let the caller compute the index themselves by computing length-1. We might need additional interfaces for that as choices array are now private.

Reply:
I see it is used in 1 test as follows:
```
	for c := 0; c < numChoices; c++ {
		ret := mab.NewChoice()
		if ret != c {
			t.Fatalf("error adding new choice %v. got return value %v", c, ret)
		}
		... use mab.choices[ret] ...
```

The test is in the same package and accesses mab.choices and assumes the returned value is the index anyways. So it can be rewritten as follows w/o losing much of the checking (or may be making it better, as we now also check exact mab.choices size):

```
	for c := 0; c < numChoices; c++ {
		mab.NewChoice()
		if len(mab.choices) != c + 1 {
			t.Fatalf("error adding new choice %v. got return value %v", c, ret)
		}
		... use mab.choices[c] ...
```